src/conic_form.jl

```diff
@@ -0,0 +1,263 @@
+end
+function MOI.set(model::ConicForm, ::MOI.ConstraintDualStart,
+                  ci::MOI.ConstraintIndex, value)
+    offset = constroffset(model, ci)
+    model.dual[rows(model, ci)] .= value
+end
+function MOI.set(model::ConicForm, ::MOI.ObjectiveSense, sense::MOI.OptimizationSense)
+    model.sense = sense
+end
+variable_index_value(t::MOI.ScalarAffineTerm) = t.variable_index.value
+variable_index_value(t::MOI.VectorAffineTerm) = variable_index_value(t.scalar_term)
+function MOI.set(model::ConicForm, ::MOI.ObjectiveFunction,
+                 f::MOI.ScalarAffineFunction{Float64})
+    c = Vector(sparsevec(variable_index_value.(f.terms), MOI.coefficient.(f.terms),
+                         model.A.n))
+    model.objective_constant = f.constant
+    model.c = c
+    return nothing
+end
+
+function _allocate_constraint(model::ConicForm, src, indexmap, cone_id, ci)
+    # TODO use `CanonicalConstraintFunction`
+    func = MOI.get(src, MOI.ConstraintFunction(), ci)
+    func = MOIU.is_canonical(func) ? func : MOI.Utilities.canonical(func)
+    allocate_terms(model.A, indexmap, func)
+    offset = model.num_rows[cone_id]
+    model.num_rows[cone_id] = offset + MOI.output_dimension(func)
+    return ci, offset, func
+end
+
+function _allocate_constraints(model::ConicForm, src, indexmap, cone_id, ::Type{S}) where S
+    cis = MOI.get(src, MOI.ListOfConstraintIndices{MOI.VectorAffineFunction{Float64}, S}())
+    return map(cis) do ci
+        _allocate_constraint(model, src, indexmap, cone_id, ci)
+    end
+end
+
+function _load_variables(model::ConicForm, nvars::Integer)
+    m = sum(model.num_rows)
+    model.A.m = m
+    model.b = zeros(m)
+    model.c = zeros(model.A.n)
+    allocate_nonzeros(model.A)
+end
+
+function _load_constraints(model::ConicForm, src, indexmap, cone_offset, i, cache, preprocess)
+    for (ci_src, offset_in_cone, func) in cache
+        offset = cone_offset + offset_in_cone
+        set = MOI.get(src, MOI.ConstraintSet(), ci_src)
+        new_func = preprocess(func, set)
+        load_terms(model.A, indexmap, new_func, offset)
+        copyto!(model.b, offset + 1, new_func.constants)
+        model.dimension[offset] = MOI.output_dimension(func)
+        indexmap[ci_src] = typeof(ci_src)(offset)
+    end
+end
+
+# Vectorized length for matrix dimension n
+sympackedlen(n) = div(n*(n+1), 2)
+# Matrix dimension for vectorized length n
+sympackeddim(n) = div(isqrt(1+8n) - 1, 2)
+function _sympackedto(x, n, mapfrom, mapto)
+    @assert length(x) == sympackedlen(n)
+    y = similar(x)
+    for i in 1:n, j in 1:i
+        y[mapto(i, j)] = x[mapfrom(i, j)]
+    end
+    y
+end
+trimap(i::Integer, j::Integer) = i < j ? trimap(j, i) : div((i-1)*i, 2) + j
+trimapL(i::Integer, j::Integer, n::Integer) = i < j ? trimapL(j, i, n) : i + div((2n-j) * (j-1), 2)
+sympackedLtoU(x, n=sympackeddim(length(x))) = _sympackedto(x, n, (i, j) -> trimapL(i, j, n), trimap)
+sympackedUtoL(x, n) = _sympackedto(x, n, trimap, (i, j) -> trimapL(i, j, n))
```

Comment:
Why are these here? This is specific to SCS needing to scale off diagonal entries, do you need this too in DiffOpt?

Reply:
sorry for the late response.
yes, we need at least this part to scale diagonal entries for a PSD cone. my only source of cross checking the actual values was https://github.com/cvxgrp/diffcp, which accepts/returns matrices in SCS format

Reply:
It doesn't make much sense to do it by default for the conic form. I would suggest DiffOpt to use something like
https://github.com/jump-dev/SCS.jl/blob/c1e3a8d5469e9f289dd0b3cdd87245792d266317/src/MOI_wrapper.jl#L147
In the future, if the scaled psd cone is brought back then SCS and DiffOpt could use this without any preprocessing function.

Reply:
See https://github.com/jump-dev/MathOptInterface.jl/issues/531 for the scaled psd cone

Reply:
ok got it now. will need to correct the tests accordingly!
should I check the presence of SCS/Mosek solver and edit the matrices accordingly - so that for every solver matrix A is same

Reply:
> should I check the presence of SCS/Mosek solver and edit the matrices accordingly - so that for every solver matrix A is same

Not sure to understand what you mean

Reply:
forgot to edit the comment 😅 . saw the conic solver info at https://github.com/cvxgrp/cvxpy/tree/master/cvxpy/reductions/solvers/conic_solvers and understood this `trimap` thing is specific to SCS and shouldn't be a part of MatOI

Reply:
@blegat is this PR good to go? or is there anything specific to SCS that I'm missing